syntax = "proto3";

package gfl.core;

enum JobStatusEnum {
  NEW = 0;
  READY = 1;
  RUNNING = 2;
  WAITING = 3;
  FINISHED = 4;
  ABORTED = 5;
}

enum DatasetType {
  IMAGE = 0;
  VIDEO = 1;
  AUDIO = 2;
  TEXT = 3;
  STRUCTURE = 4;
}

message JobMeta {
  string id = 1;
  string owner = 2;
  int64 create_time = 3;
  string content = 4;
  JobStatusEnum status = 5;
  repeated string dataset_ids = 6;
}

message JobConfig {
  string trainer = 1;
  string aggregator = 2;
}

message TrainConfig {
  string model = 1;
  string optimizer = 2;
  string criterion = 3;
  string lr_scheduler = 4;
  int32 epoch = 5;
  int32 batch_size = 6;
}

message AggregateConfig {
  int32 global_epoch = 1;
}

message JobData {
  JobMeta meta = 1;
  JobConfig job_config = 2;
  TrainConfig train_config = 3;
  AggregateConfig aggregate_config = 4;
}

message DatasetMeta {
  string id = 1;
  string owner = 2;
  int64 create_time = 3;
  string content = 4;
  DatasetType type = 5;
  int32 size = 6;
}

message DatasetConfig {
  string dataset = 1;
  string val_dataset = 2;
  float val_rate = 3;
}

message DatasetData {
  DatasetMeta meta = 1;
  DatasetConfig dataset_config = 2;
}

message ModelParams {
  string job_id = 1;
  string node_id = 2;
  int32 step = 3;
  bytes data = 4;
}